src/storage/src/sst/stream_writer.rs

```diff
@@ -0,0 +1,196 @@
+        Self {
+            buffer: Arc::new(Mutex::new(BytesMut::with_capacity(size))),
+        }
+    }
+}
+
+impl Write for Buffer {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        let len = buf.len();
+        let mut buffer = self.buffer.lock().unwrap();
+        buffer.put_slice(buf);
+        Ok(len)
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        Ok(())
+    }
+}
+
+/// Parquet writer that buffers row groups in memory and writes buffered data to an underlying
+/// storage by chunks to reduce memory consumption.
+pub struct BufferedWriter {
+    path: String,
+    arrow_writer: ArrowWriter<Buffer>,
+    object_writer: Writer,
+    buffer: Buffer,
+    bytes_written: AtomicU64,
+    flushed: bool,
+    threshold: usize,
+    arrow_schema: arrow::datatypes::SchemaRef,
+}
+
+impl BufferedWriter {
+    pub async fn try_new(
+        path: String,
+        store: ObjectStore,
+        schema: &SchemaRef,
+        props: Option<WriterProperties>,
+        buffer_threshold: usize,
+    ) -> error::Result<Self> {
+        let arrow_schema = schema.arrow_schema();
+        let buffer = Buffer::with_capacity(buffer_threshold);
+        let writer = store
+            .writer(&path)
+            .await
+            .context(WriteObjectSnafu { path: &path })?;
+
+        let arrow_writer = ArrowWriter::try_new(buffer.clone(), arrow_schema.clone(), props)
+            .context(WriteParquetSnafu)?;
+
+        Ok(Self {
+            path,
+            arrow_writer,
+            object_writer: writer,
+            buffer,
+            bytes_written: Default::default(),
+            flushed: false,
+            threshold: buffer_threshold,
+            arrow_schema: arrow_schema.clone(),
+        })
+    }
+
+    /// Write a record batch to stream writer.
+    pub async fn write(&mut self, batch: &Batch) -> error::Result<()> {
+        let arrow_batch = RecordBatch::try_new(
+            self.arrow_schema.clone(),
+            batch
+                .columns()
+                .iter()
+                .map(|v| v.to_arrow_array())
+                .collect::<Vec<_>>(),
+        )
+        .context(NewRecordBatchSnafu)?;
+        self.arrow_writer
+            .write(&arrow_batch)
+            .context(WriteParquetSnafu)?;
+        let written = Self::try_flush(
+            &self.path,
+            &self.buffer,
+            &mut self.object_writer,
+            false,
+            &mut self.flushed,
+            self.threshold,
+        )
+        .await?;
+        self.bytes_written.fetch_add(written, Ordering::Relaxed);
+        Ok(())
+    }
+
+    /// Abort writer.
+    pub async fn abort(self) -> bool {
+        // TODO(hl): Currently we can do nothing if file's parts have been uploaded to remote storage
+        // on abortion, we need to find a way to abort the upload. see https://help.aliyun.com/document_detail/31996.htm?spm=a2c4g.11186623.0.0.3eb42cb7b2mwUz#reference-txp-bvx-wdb
```

Comment:
Can we just delete the file in remote?

Reply:
> Can we just delete the file in remote?

When using multipart upload, there's no visible partial file unless the upload is completed. We need to keep the upload id an manually abort the upload.